src/layers/core/path-layer/path-layer.js

```diff
@@ -153,69 +135,72 @@ export default class PathLayer extends Layer {
-        positions[i++] = point[0];
-        positions[i++] = point[1];
-        positions[i++] = point[2] || 0;
       })
     );
+    positions.copyWithin(i, 0, i);
 
     attribute.value = positions;
   }
 
-  calculatePrevPositions(attribute) {
+  calculateLeftDeltas(attribute) {
     const {paths, pointCount} = this.state;
-    const prevPositions = new Float32Array(pointCount * attribute.size * 2);
+
+    const leftDeltas = new Float32Array(pointCount * attribute.size * 2);
 
     let i = 0;
     paths.forEach(path => {
-      this._shiftPath(path, -1).forEach(point => {
-        prevPositions[i++] = point[0];
-        prevPositions[i++] = point[1];
-        prevPositions[i++] = point[2] || 0;
-        prevPositions[i++] = point[0];
-        prevPositions[i++] = point[1];
-        prevPositions[i++] = point[2] || 0;
-      });
+      i += 3;
+      path.reduce((prevPoint, point) => {
+        if (prevPoint) {
+          leftDeltas[i++] = point[0] - prevPoint[0];
+          leftDeltas[i++] = point[1] - prevPoint[1];
+          leftDeltas[i++] = (point[2] - prevPoint[2]) || 0;
+        }
+        return point;
+      }, null);
     });
+    leftDeltas.copyWithin(i, 0, i);
 
-    attribute.value = prevPositions;
+    attribute.value = leftDeltas;
   }
 
-  calculateNextPositions(attribute) {
+  calculateRightDeltas(attribute) {
     const {paths, pointCount} = this.state;
-    const nextPositions = new Float32Array(pointCount * attribute.size * 2);
+
+    const rightDeltas = new Float32Array(pointCount * attribute.size * 2);
 
     let i = 0;
     paths.forEach(path => {
-      this._shiftPath(path, 1).forEach(point => {
-        nextPositions[i++] = point[0];
-        nextPositions[i++] = point[1];
-        nextPositions[i++] = point[2] || 0;
-        nextPositions[i++] = point[0];
-        nextPositions[i++] = point[1];
-        nextPositions[i++] = point[2] || 0;
-      });
+      path.reduce((prevPoint, point) => {
+        if (prevPoint) {
+          rightDeltas[i++] = point[0] - prevPoint[0];
+          rightDeltas[i++] = point[1] - prevPoint[1];
+          rightDeltas[i++] = (point[2] - prevPoint[2]) || 0;
+        }
+        return point;
+      }, null);
+      i += 3;
     });
+    rightDeltas.copyWithin(i, 0, i);
 
-    attribute.value = nextPositions;
+    attribute.value = rightDeltas;
   }
 
   calculateDirections(attribute) {
     const {data, getWidth} = this.props;
     const {paths, pointCount} = this.state;
-    const directions = new Float32Array(pointCount * attribute.size * 2);
+    const directions = new Float32Array(pointCount * 2);
 
     let i = 0;
     paths.forEach((path, index) => {
       let w = getWidth(data[index], index);
       if (isNaN(w)) {
         w = 1;
       }
-
-      path.forEach(() => {
-        directions[i++] = w;
-        directions[i++] = -w;
-      });
+      const count = path.length;
+      fillArray({target: directions, source: [w], start: i, count});
```

Comment:
If it is just one element, `Array.fill` might be faster (and standard JS). `fillArray` was written to quickly duplicate a subarray.